modules/internal/ChapelArray.chpl

```diff
@@ -1733,15 +1744,80 @@ module ChapelArray {
        instance of ``val`` in the array, or if ``val`` is not found, a
        tuple containing ``false`` and an unspecified value is returned.
      */
-     @unstable "'Array.find' is unstable"
-     proc find(val: this.eltType): (bool, index(this.domain)) {
+     deprecated "The tuple-returning version of `.find() on arrays` is deprecated; to opt into the new index-returning version, recompile with `-suseNewArrayFind`"
+     proc find(val: this.eltType): (bool, index(this.domain)) where !useNewArrayFind {
       for i in this.domain {
         if this[i] == val then return (true, i);
       }
       var arbInd: index(this.domain);
       return (false, arbInd);
     }
 
+
+    /*
+
+      Search an array for ``val``, returning whether or not it is
+      found.  If the value is found, the index storing it is returned
+      in ``idx``.  If it is not found, the final value of ``idx`` is
+      unspecified.
+
+    */
+    proc find(val: eltType, ref idx: fullIdxType): bool {
+      // For the sparse case, start by seeing if the IRV is what we're
+      // looking for.  If so, iterate over the parent domain to look
+      // for the value or an index not represented in the array.  This
+      // assumes that we're likely to find a case of the IRV fast.
+      // Otherwise, fall through to the normal case to just search the
+      // explicitly represented values.
+      if this.isSparse() && val == this.IRV {
+        for i in this.domain._value.parentDom {
+          // If we find an index representing the IRV, return it; if
+          // it has an explicit value, it may still be the IRV, so
+          // check it.
+          if !this.domain.contains(i) || this[i] == val {
+            idx = i;
+            return true;
+          }
+        }
+      } else {
+        for i in this.domain {
+          if this[i] == val {
+            idx = i;
+            return true;
+          }
+        }
+      }
+
+      // We didn't find it, so return false.
+      return false;
+    }
+
+    /*
+
+      Search a rectangular array with integral indices for ``val``,
+      returning the index where it is found.  If the array contains
+      multiple copies of ``val``, the lexicographically earliest index
+      will be returned.  If ``val`` is not found,
+      ``domain.lowBound-1`` will be returned instead.
+
+      Note that for arrays with ``idxType=int(?w)`` (signed ``int``
```

Comment:
This would be an issue for arrays with unsigned integers too if the low bound was 0 right? I think I'd just generalize the statement to say if the low bound in a dimension is min(idxType) the result is not well-defined.

Reply:
I don't think so.  My understanding is that unsigned integer wraparound is well-defined (so the result will be `max(uint(w))`.  Rather than spelling that out explicitly, I was hoping we could rely on the user understanding (or figuring out) that `0-1` would be `max(uint(w))` for a uint idxType.